src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -1129,61 +1129,38 @@ public ConnectionState(ConnectionContext connection, HubConnection hubConnection
 
             public void AddInvocation(InvocationRequest irq)
             {
-                lock (_lock)
+                if (!_pendingCalls.TryAdd(irq.InvocationId, irq))
                 {
-                    if (_pendingCalls.ContainsKey(irq.InvocationId))
-                    {
-                        Log.InvocationAlreadyInUse(_hubConnection._logger, irq.InvocationId);
-                        throw new InvalidOperationException($"Invocation ID '{irq.InvocationId}' is already in use.");
-                    }
-                    else
-                    {
-                        _pendingCalls.Add(irq.InvocationId, irq);
-                    }
+                    Log.InvocationAlreadyInUse(_hubConnection._logger, irq.InvocationId);
+                    throw new InvalidOperationException($"Invocation ID '{irq.InvocationId}' is already in use.");
                 }
             }
 
-            public bool TryGetInvocation(string invocationId, out InvocationRequest irq)
-            {
-                lock (_lock)
-                {
-                    return _pendingCalls.TryGetValue(invocationId, out irq);
-                }
-            }
+            public bool TryGetInvocation(string invocationId, out InvocationRequest irq) => _pendingCalls.TryGetValue(invocationId, out irq);
+
+            public bool TryRemoveInvocation(string invocationId, out InvocationRequest irq) => _pendingCalls.TryRemove(invocationId, out irq);
 
-            public bool TryRemoveInvocation(string invocationId, out InvocationRequest irq)
-            {
-                lock (_lock)
-                {
-                    if (_pendingCalls.TryGetValue(invocationId, out irq))
-                    {
-                        _pendingCalls.Remove(invocationId);
-                        return true;
-                    }
-                    else
-                    {
-                        return false;
-                    }
-                }
-            }
 
             public void CancelOutstandingInvocations(Exception exception)
             {
                 Log.CancelingOutstandingInvocations(_hubConnection._logger);
 
-                lock (_lock)
+                // https://msdn.microsoft.com/en-us/library/dd287131.aspx
```

Comment:
Lets go ahead and remove this 👍

Reply:
The behavior here has slightly changed. Even though it's safe to enumerate the dictionary itself in a lock, it was being synchronized with other operations (like StopAsync).

Reply:
I will try to demonstrate what you are saying, but the scope for that unit tests  is not very simple because that is a inner private class.  
Some questions regarding the usage of ConnectionState: 
An instance of ConnectionState exists per connection that lives in a Lock Sync of one unit ( method StartAsyncCore of HubConnection class ) that means that you will have only one thread executing that scope of code at a time no ? So from where will it come more threads that may be executing that code concurrently ?